File: php/20-valid-parentheses.php
<?php
class Solution 
{
    /**
     * @param String $s
     * @return Boolean
     */
    public function isValid($s) 
    {
        $stack = [];
        $strlen = strlen($s);
        if ($strlen == 0) {
            return true;
        }
        $left = ['('=>0, '{'=>1, '['=>2];
        $right = [')'=>0, '}'=>1, ']'=>2];
        
        for ($i = 0; $i < $strlen; $i++) {
            $str = $s[$i];
            if (!isset($left[$str) && !isset($right[$str])) {
                return false;
            }
            if (isset($left[$str])) {
                array_push($stack, $str);
            } else {
                $c = array_pop($stack);
                if ($left[$c] !== $right[$str]) {
                    return false;
                }
            }
        }
        return empty($stack)? true : false;
    }
}

File: go/20-valid-parentheses.go
package main

func isValid(s string) bool {
	if len(s) == 0 {
		return true
	}
	left := map[string]int{
		"(": 0,
		"[": 1,
		"{": 2,
	}
	right := map[string]int{
		")": 0,
		"]": 1,
		"}": 2,
	}
	stack := []string{}

	for _, c := range s {
		p := string(c)
		_, inLeft := left[p]
		_, inRight := right[p]
		if !inLeft && !inRight {
			return false
		}
		if inLeft {
			stack = append(stack, p)
		} else {
			if len(stack) <= 0 {
				return false
			}
			match := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if left[match] != right[p] {
				return false
			}
		}
	}
	if len(stack) > 0 {
		return false
	}
	return true
}
